Source/WebCore/Modules/screen-wake-lock/WakeLock.cpp

```diff
@@ -26,16 +26,72 @@
+WakeLock::WakeLock(Document* document)
+    : ContextDestructionObserver(document)
+{
+}
+
+// https://www.w3.org/TR/screen-wake-lock/#the-request-method
+void WakeLock::request(WakeLockType lockType, Ref<DeferredPromise>&& promise)
+{
+    auto* document = this->document();
+    if (!document || !document->isFullyActive() || !document->page()) {
+        promise->reject(Exception { NotAllowedError, "Document is not fully active"_s });
+        return;
+    }
+    if (!isFeaturePolicyAllowedByDocumentAndAllOwners(FeaturePolicy::Type::ScreenWakeLock, *document, LogFeaturePolicyFailure::Yes)) {
+        promise->reject(Exception { NotAllowedError, "'screen-wake-lock' is not allowed by Feature-Policy"_s });
+        return;
+    }
+    if (document->visibilityState() == VisibilityState::Hidden) {
+        promise->reject(Exception { NotAllowedError, "Document is hidden"_s });
+        return;
+    }
+
+    // FIXME: The permission check can likely be dropped once the specification gets updated to only
+    // require transient activation (https://github.com/w3c/screen-wake-lock/pull/326).
+    bool hasTransientActivation = document->domWindow() && document->domWindow()->hasTransientActivation();
+    PermissionController::shared().query(document->clientOrigin(), PermissionDescriptor { PermissionName::ScreenWakeLock }, *document->page(), PermissionQuerySource::Window, [protectedThis = Ref { *this }, document = Ref { *document }, hasTransientActivation, promise = WTFMove(promise), lockType](std::optional<PermissionState> permission) mutable {
+        if (!permission || *permission == PermissionState::Prompt)
+            permission = hasTransientActivation ? PermissionState::Granted : PermissionState::Denied;
+        document->eventLoop().queueTask(TaskSource::ScreenWakelock, [protectedThis = WTFMove(protectedThis), document = WTFMove(document), promise = WTFMove(promise), lockType, permission]() mutable {
+            if (permission == PermissionState::Denied) {
+                promise->reject(Exception { NotAllowedError, "Permission was denied"_s });
+                return;
+            }
+            if (!document->isFullyActive()) {
+                promise->reject(Exception { NotAllowedError, "Document is not fully active"_s });
+                return;
+            }
+            if (document->visibilityState() == VisibilityState::Hidden) {
+                promise->reject(Exception { NotAllowedError, "Document is hidden"_s });
+                return;
+            }
+            auto lock = WakeLockSentinel::create(document, lockType);
+            promise->resolve<IDLInterface<WakeLockSentinel>>(lock.get());
+            document->wakeLockManager().addWakeLock(WTFMove(lock));
+        });
+    });
+}
 
-void WakeLock::request(WakeLockType, Ref<DeferredPromise>&& promise)
+Document* WakeLock::document()
 {
-    promise->reject(Exception { NotSupportedError });
+    return downcast<Document>(scriptExecutionContext());
```

Comment:
To be conservative, could we return null if it's not a Document?  (just in case someone in the future adds this to workers)

Reply:
But the type is not wrong. The WakeLock constructor takes a Document (because WakeLock is not exposed to workers) and thus its script execution context is always a Document.